.github/CONTRIBUTING.md

```diff
@@ -0,0 +1,89 @@
+
+# Commits
+Besides the changes themselves, commits are the most important part of contributions. There are 2 major things to keep in mind for commits:
+
+1. [Scale/Scope](#scale-and-scope)
+2. [Messages](#messages)
+
+## Scale and Scope
+The scale and scope of a commit should be reasonable. Do not commit for every line when making multiple changes, for example. However you should not include many unrelated changes in a single commit. By limiting the scope of the commit we can ensure that if any regressions or new bugs are introduced we can easily revert those changes without the need for major refactors or reimplementations. Limiting the scale of commits also makes review of the changes easier and faster.
+
+## Messages
+Commit messages should adequately explain the changes in the commit. Messages like "Updated file.md" and "spelling error" should not be used. Nonsense messages such as "oops" or "fixed" are especially not allowed. Commit messages should, at minimum, be in the format `type: message` where `type` represents the type or scope of the changes (`chore`, `docs`, `bug-fix`, etc.) and `message` is the actual changes. Unless the word is from the codebase and starts with a capital letter (such as an exported Go struct), the `message` should be lowercase. We also recommend using both "subject" and "body" commits. This can be achieved through the git CLI by using multiple `-m`/`--message` flags. For example `git commit -m "short subject" -m "longer description of the changes"`.
+
+The following are examples of good commit messages:
+
+- `chore: renamed nnid service to nnas`
+- `bug-fix: fixed hang in MutexMap.Has`
+
+Please refer to [Conventional Commits](https://conventionalcommits.org/) for a detailed guide on how to structure commit messages. Writing good, detailed, commit messages helps ensure that we can refer back to the git history and quickly find where specific changes occurred in the event that they need further review, reverting, etc.
+
+# Pull Requests
+As stated in [Making Issues](#making-issues), before making a pull request there must be an open, approved, issue for the changes being made. If you can not find an approved issue for the changes you wish to make, please make one before continuing.
+
+Before making a pull request ensure you have tested all changes and that no regressions have been introduced.
+
+Pull requests should never be made against the default (`main`/`master`) branch of a repository. The default branch contains the most recent, stable, version of the codebase. All work on the codebase should take place in other branches. Unless otherwise specified, your target branch should typically be the `dev` branch. You may target other feature branches, however, if need be. If a `dev` branch does not exist for the repository you are working on, please submit a feature request for one to be added before continuing.
```

Comment:
I don't have `dev` branches on Inkay at least, we all commit to `main` and make releases for the stable build. Should I learn a new workflow? Can this be enforced via branch protection/PR templates somehow?

Reply:
Making a `dev` branch on all the repos is something we've been talking about for quite some time now (like a year or so). So this was going to be brought up either way

Your flow wouldn't really change all that much. Just working in `dev` rather than `main` and when a release is ready to be made merging and making the release. It's almost identical just with an extra step

> Can this be enforced via branch protection/PR templates somehow?

I may be stupid, wdym?

Reply:
Like, forcing PRs to be against dev rather than main.

Reply:
Yes this can be done via branch protection. There's already a handful of repositories in the org which protect the default branch from direct commits and require PRs, so that would be added as well

The point about targeting the dev branches was there not only to tell people not to even try and PR against the default branch, but to also not base their changes off of it (which isn't something we can directly control. The most we can do is tell them not to in a document like this, but we can't physically stop someone from making a branch to make changes based off the default branch locally)